# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [2]:
import pandas as pd
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
df = pd.read_csv(url)

In [3]:
df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [4]:
df.tail()

Customer   ST GENDER Education Customer Lifetime Value  Income  \
4003      NaN  NaN    NaN       NaN                     NaN     NaN   
4004      NaN  NaN    NaN       NaN                     NaN     NaN   
4005      NaN  NaN    NaN       NaN                     NaN     NaN   
4006      NaN  NaN    NaN       NaN                     NaN     NaN   
4007      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
4003                   NaN                       NaN         NaN   
4004                   NaN                       NaN         NaN   
4005                   NaN                       NaN         NaN   
4006                   NaN                       NaN         NaN   
4007                   NaN                       NaN         NaN   

     Vehicle Class  Total Claim Amount  
4003           NaN                 NaN  
4004           NaN                 NaN  
4005           NaN                 NaN  
4006           NaN                 NaN  
4007           NaN                 NaN

In [5]:
df.shape
#This dataframe has a total of 4008 rows and 11 columns

(4008, 11)

In [6]:
df.dtypes
#The Customer Lifetime Value should be a float 

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [7]:
df['Customer Lifetime Value'] = df['Customer Lifetime Value'].str.replace('%', '').astype(float) / 100
#Cleaning the Customer Lifetime Value column and checking if it was correctly cleaned
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value      float64
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [8]:
df.nunique()
#Identifying the number of unique values for each category

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [9]:
df.columns = df.columns.str.title()

In [10]:
categorical_cols = df.select_dtypes(include='object').head()
categorical_cols
#Identifying which columns appear to be categorical

Customer          St Gender             Education Number Of Open Complaints  \
0  RB50392  Washington    NaN                Master                    1/0/00   
1  QZ44356     Arizona      F              Bachelor                    1/0/00   
2  AI49188      Nevada      F              Bachelor                    1/0/00   
3  WW63253  California      M              Bachelor                    1/0/00   
4  GA49547  Washington      M  High School or Below                    1/0/00   

      Policy Type  Vehicle Class  
0   Personal Auto  Four-Door Car  
1   Personal Auto  Four-Door Car  
2   Personal Auto   Two-Door Car  
3  Corporate Auto            SUV  
4   Personal Auto  Four-Door Car

In [11]:
for column in categorical_cols:
    print(f"Unique values for the {column} column:")
    print(df[column].unique(), '\n')

Unique values for the Customer column:
['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan] 

Unique values for the St column:
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan] 

Unique values for the Gender column:
[nan 'F' 'M' 'Femal' 'Male' 'female'] 

Unique values for the Education column:
['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan] 

Unique values for the Number Of Open Complaints column:
['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan] 

Unique values for the Policy Type column:
['Personal Auto' 'Corporate Auto' 'Special Auto' nan] 

Unique values for the Vehicle Class column:
['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan] 



#### Cleaning each categorical column

In [12]:
numerical_cols = round(df.select_dtypes(include='number').head(), 2)
numerical_cols
#Identifying which columns appear to be numerical

Customer Lifetime Value   Income  Monthly Premium Auto  Total Claim Amount
0                      NaN      0.0                1000.0                2.70
1                  6979.54      0.0                  94.0             1131.46
2                 12887.43  48767.0                 108.0              566.47
3                  7645.86      0.0                 106.0              529.88
4                  5363.08  36357.0                  68.0               17.27

In [13]:
round(numerical_cols.describe().loc[['min', 'max']], 2)

Customer Lifetime Value   Income  Monthly Premium Auto  \
min                  5363.08      0.0                  68.0   
max                 12887.43  48767.0                1000.0   

     Total Claim Amount  
min                2.70  
max             1131.46

In [14]:
#The min and max Customer Lifetime Value indicate the lowest and highest revenue a customer will generate for their insurance company. The large gap between those indicates that there is a large variance in Customer Lifetime Value.
#The min and max Income indicate that there is a wide range of customer financial profiles. This might indicate that there are customers with no income at all or this data could be missing.
#The min and max Monthly Premium Auto indicates that there is a wide range in terms of coverage. There are many factors that can contribute to this such as: the coverage a customer wants for their car, vehicle type, total claim amounts.
#The min and max Total Claim Amount indicates that some customers have had significant claims while others have made few or no claims. The high claim amount might indicate a higher amount of claims or a big incident.

In [15]:
summary_statistics = round(numerical_cols.describe(), 2)
summary_statistics

Customer Lifetime Value    Income  Monthly Premium Auto  \
count                     4.00      5.00                  5.00   
mean                   8218.98  17024.80                275.20   
std                    3256.55  23721.47                405.49   
min                    5363.08      0.00                 68.00   
25%                    6575.42      0.00                 94.00   
50%                    7312.70      0.00                106.00   
75%                    8956.25  36357.00                108.00   
max                   12887.43  48767.00               1000.00   

       Total Claim Amount  
count                5.00  
mean               449.56  
std                466.82  
min                  2.70  
25%                 17.27  
50%                529.88  
75%                566.47  
max               1131.46

- On average, a Customer Lifetime Value is 8219. Since the median is slightly lower than the mean it indicates that some higher values might be pulling up the mean. The STD gives us the indication that there is a high spread of data. 75% of customers have a CLV above 8956 while 50% of them have a Customer Lifetime Value above 7313,  and 25% have a CLV below 6575.
- The average income is 17025. The fact that the median is zero suggests that there some customers have no reported income or that this data is missing. We can also see this by looking at Q2, 50% of the customers have no recorded income. The STD is very high which indicates that there is a high range when it comes to the income, some customers earn vey low while others are reported to earn significantly more.
- On average customers pay 275 for their insurance. The medican is lower than the mean which suggests that there are a couple os customers are paying very high insurance, making the average go up. The STD is 405.49 (not close to the mean) which means that some customers are paying significantly lower insurance while other are paying significantly high insurance. Since Q1 is 94 and Q3 is 108, it means that most customers are paying between these values.
- On average customers have made claims totaling 449.56. The median is higher than the mean which means that a couple of customers have made large claims. The STD indicates that there is a wide range in terms of claim amounts. Since Q1 is 17.27 and Q3 is 566.47 it means that most customers fall in between these values.

In [16]:
df.describe(include='object')

Customer      St Gender Education Number Of Open Complaints  \
count      1071    1071    954      1071                      1071   
unique     1071       8      5         6                         6   
top     MY31220  Oregon      F  Bachelor                    1/0/00   
freq          1     320    457       324                       830   

          Policy Type  Vehicle Class  
count            1071           1071  
unique              3              6  
top     Personal Auto  Four-Door Car  
freq              780            576

- There are 1071 unique IDs which indicates that most IDs are null values or missing and that there are no duplicates.
- Every row has the State listed, there are 8 total states but as we've seen before this is not 100% correct as we have the same states written differently. The most frequent is Oregon with a frequency of 320. This suggests that most customers are from Oregon.
- When it comes to gender we're able to observe that there are some missing values. The 5 unique values indicates that the Genders have also been written differently. Among these entries we can atest that 457 are females but we'll be able to see this in a clearer way after cleaning the data.
- When it comes to Education we can observe that Bachelor's is the most common education level.
- The Number of Open Complains is not clear, there is no clear indication of what 1/0/00 means.
- The Policy Type shows that the majority of the customers have a Personal Car.
- The Vehicle Class shows that there are 6 vehicle class types but the most common vehicle is a Four-Door-Car.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [17]:
customer_locations = pd.read_csv(url, usecols=["ST"]).squeeze("columns")
customer_locations

0       Washington
1          Arizona
2           Nevada
3       California
4       Washington
           ...    
4003           NaN
4004           NaN
4005           NaN
4006           NaN
4007           NaN
Name: ST, Length: 4008, dtype: object

In [18]:
customer_locations.unique()

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', 'Cali',
       'AZ', 'WA', nan], dtype=object)

In [19]:
customer_locations.replace({'AZ': 'Arizona', 'WA': 'Washington', 'Cali': 'California'}, inplace=True)

In [20]:
customer_locations.nunique()

5

In [21]:
customer_locations.value_counts().sort_values().head(5)

ST
Nevada         98
Washington    111
Arizona       211
Oregon        320
California    331
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [22]:
policy_type = pd.read_csv(url, usecols=["Policy Type"]).squeeze("columns")
policy_type

0        Personal Auto
1        Personal Auto
2        Personal Auto
3       Corporate Auto
4        Personal Auto
             ...      
4003               NaN
4004               NaN
4005               NaN
4006               NaN
4007               NaN
Name: Policy Type, Length: 4008, dtype: object

In [23]:
policy_type.value_counts()

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [24]:
policy_type.value_counts().idxmax()

'Personal Auto'

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [25]:
personal_auto = df.loc[df['Policy Type'] == 'Personal Auto']
corporate_auto = df.loc[df['Policy Type'] == 'Corporate Auto']

In [26]:
avg_personal_auto = personal_auto['Income'].mean()
avg_corporate_auto = corporate_auto['Income'].mean()

print(avg_personal_auto)
print(avg_corporate_auto)

38180.69871794872
41390.31196581197


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [27]:
df['Total Claim Amount'].describe()

count    1071.000000
mean      404.986909
std       293.027260
min         0.382107
25%       202.157702
50%       354.729129
75%       532.800000
max      2893.239678
Name: Total Claim Amount, dtype: float64

In [28]:
total_claim_75 = df['Total Claim Amount'].quantile(0.75)

In [30]:
high_policy_claim = df[df['Total Claim Amount'] >= total_claim_75]

In [31]:
high_policy_claim.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count               273.000000    273.000000            273.000000   
mean               9183.794161  24106.545788            163.000000   
std                6582.581975  27542.717560            613.645376   
min                2287.596900      0.000000             63.000000   
25%                4747.734600      0.000000             99.000000   
50%                7859.414600  18846.000000            113.000000   
75%               10656.882000  44897.000000            133.000000   
max               40224.013600  99316.000000          10202.000000   

       Total Claim Amount  
count          273.000000  
mean           774.005353  
std            291.304731  
min            532.800000  
25%            595.200000  
50%            673.342650  
75%            844.481918  
max           2893.239678